garage/tf/models/base.py

```diff
@@ -0,0 +1,297 @@
+        else:
+            _variable_scope = tf.variable_scope(
+                self._name, reuse=True, auxiliary_name_scope=False)
+            if network_name in self._networks:
+                raise ValueError(
+                    'Network {} already exists!'.format(network_name))
+            with _variable_scope:
+                with tf.name_scope(name=network_name):
+                    network = Network()
+                    network._inputs = inputs
+                    network._outputs = self._build(*inputs)
+        spec = self.network_output_spec()
+        if spec:
+            c = namedtuple(network_name,
+                           [*spec, 'input', 'output', 'inputs', 'outputs'])
+            if isinstance(network.outputs, tuple):
+                assert len(spec) == len(network.outputs), (
+                    'network_output_spec must have same length as outputs!')
+                self._networks[network_name] = c(
+                    *network.outputs, network.input, network.output,
+                    network.inputs, network.outputs)
+            else:
+                self._networks[network_name] = c(
+                    network.outputs, network.input, network.output,
+                    network.inputs, network.outputs)
+        else:
+            self._networks[network_name] = network
+        return network.outputs
+
+    def _build(self, *inputs):
+        """
+        Output of the model given input placeholder(s).
+
+        User should implement _build() inside their subclassed model,
+        and construct the computation graphs in this function.
+
+        Args:
+            inputs: Tensor input(s), recommended to be position arguments, e.g.
+              def build(self, state_input=None, action_input=None, name=None).
+              It would be usually same as the inputs in build().
+        Return:
+            output: Tensor output(s) of the model.
+        """
+        pass
+
+    def network_output_spec(self):
+        """
+        Network output spec.
+
+        Return:
+            *inputs: List of key(str) for the network outputs.
+        """
+        return []
+
+    @property
+    def networks(self):
+        """Networks of the model."""
+        return self._networks
+
+    @property
+    def parameters(self):
+        """Parameters of the model."""
+        return tf.get_default_session().run(self._get_variables())
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set model parameters."""
+        variables = self._get_variables()
+        for name, var in variables.items():
+            if name in parameters:
+                var.load(parameters[name])
+            else:
+                warnings.warn('No value provided for variable {}'.format(name))
+
+    @property
+    def name(self):
+        """Name of the model."""
+        return self._name
+
+    def _get_variables(self):
+        return {v.name: v for v in tf.get_variable_scope().global_variables()}
```

Comment:
don't use `tf.get_variable_scope()`, which points to the current scope in this thread regardless of what scope the model uses. add a test case to provide it to yourself.

instead, store the `Scope` object and then call `self._variable_scope.global_variables()`

Reply:
a `variable_scope` doesn't have `global_variables()`, only a `VariableScope` does, but we can only enter a `variable_scope`, not `VariableScope`.
I will use `tf.global_variables(self._variable_scope._name_or_scope)`.

Reply:
ok